tools/bss_reordering.py

```diff
@@ -0,0 +1,207 @@
+            return []
+
+        relocs = []
+        offset_hi16 = 0
+        for reloc in reloc_section.iter_relocations():
+            reloc_offset = reloc.entry["r_offset"]
+            reloc_type = reloc.entry["r_info_type"]
+            reloc_name = symtab.get_symbol(reloc.entry["r_info_sym"]).name
+
+            if reloc_type == 2:  # R_MIPS_32
+                offset_32 = reloc_offset
+                addend = int.from_bytes(
+                    data[reloc_offset : reloc_offset + 4], "big", signed=True
+                )
+                relocs.append(Reloc(reloc_name, offset_32, None, None, addend))
+            elif reloc_type == 4:  # R_MIPS_26
+                pass
+            elif reloc_type == 5:  # R_MIPS_HI16
+                offset_hi16 = reloc_offset
+            elif reloc_type == 6:  # R_MIPS_LO16
+                offset_lo16 = reloc_offset
+                addend_hi16 = int.from_bytes(
+                    data[offset_hi16 + 2 : offset_hi16 + 4], "big", signed=False
+                )
+                addend_lo16 = int.from_bytes(
+                    data[offset_lo16 + 2 : offset_lo16 + 4], "big", signed=True
+                )
+                addend = (addend_hi16 << 16) + addend_lo16
+                relocs.append(
+                    Reloc(reloc_name, None, offset_hi16, offset_lo16, addend)
+                )
+            else:
+                raise NotImplementedError(f"Unsupported relocation type: {reloc_type}")
+
+        return relocs
+
+
+def read_u32(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(4), "big")
+
+
+def read_u16(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(2), "big")
+
+
+def read_s16(f: BinaryIO, offset: int) -> int:
+    f.seek(offset)
+    return int.from_bytes(f.read(2), "big", signed=True)
+
+
+def main():
+    parser = argparse.ArgumentParser(
+        description="Report BSS reorderings between the baserom and the current build "
+        "by parsing relocations from the built object files and comparing their final values "
+        "between the baserom and the current build. "
+        "Assumes that the only differences are due to BSS ordering."
+    )
+    parser.add_argument(
+        "--oot-version",
+        "-v",
+        type=str,
+        default="gc-eu-mq-dbg",
+        help="OOT version (default: gc-eu-mq-dbg)",
+    )
+    parser.add_argument(
+        "--segment",
+        type=str,
+        help="ROM segment to check, e.g. 'boot', 'code', or 'ovl_player_actor' (default: all)",
+    )
+
+    args = parser.parse_args()
+    version = args.oot_version
+
+    mapfile = mapfile_parser.mapfile.MapFile()
+    mapfile.readMapFile(f"build/{version}/oot-{version}.map")
+
+    mapfile_segments = []
+    for mapfile_segment in mapfile:
+        if (
+            args.segment
+            and mapfile_segment.name != f"..{args.segment}"
+            and mapfile_segment.name != f"..{args.segment}.bss"
```

Comment:
If I understand what the script is doing correctly, I think you can not put .bss sections in mapfile_segments at all

Afaict mapfile_segments is only used for getting relocs and diffing expected-built, so examining bss for either is a noop (there are never any relocs in bss) or nonsensical (you can't diff bss, it's not in rom)

EDIT:
actually mapfile_segments is also iterated over for bss sections, I'd suggest having two lists, one "segments_load" (LOAD being linker lingo for "put into the rom", feel free to find something else) and another "segments_bss"

Reply:
Iterating over the BSS sections does not take very long, so I make this list only the segments we'll be searching for relocs (now called `reloc_mapfile_segments`) and changed the BSS loop to go through every BSS segment. It might not be entirely accurate but I think filtering them again is not worth the complexity

Reply:
I think it's worth the readability but ok whatever